examples/userapi/subdomains.ipynb

```diff
@@ -0,0 +1,954 @@
+    "\n",
+    "Now, lets us compose a subdomain from our class `FullDomain`:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 11,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "full_domain = FullDomain()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "It it important to note that whilst the `full_domain` object has now been initialised, it has not yet been finalised, i.e. it is not currently associated with any grid. To attach this subdomain to a grid we must pass it to the grid during its creation (the sudomain will then be finalised during creation of the grid):"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 12,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "my_grid = Grid(shape = (10, 10, 10), subdomains = (full_domain, ))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "If we look at the subdomains of `my_grid` we see that our new subdomain 'mydomain' is now present:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 13,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "{'domain': SubDomain domain[(x, y, z)],\n",
+       " 'interior': SubDomain interior[(xi, yi, zi)],\n",
+       " 'mydomain': SubDomain mydomain[(x, y, z)]}"
+      ]
+     },
+     "execution_count": 13,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "my_grid.subdomains"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "As expected (and intended), the dimensions of our newly defined subdomain match those of 'domain'.\n",
+    "\n",
+    "Now, let us create a subdomain consisting of all but the outer grid point in each dimension (similar to 'interior'):"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 14,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class Innerdomain(SubDomain):\n",
+    "    name = 'inner'\n",
+    "    def define(self, dimensions):\n",
+    "        d = dimensions\n",
+    "        return {d: ('middle', 1, 1) for d in dimensions}"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "First, note that in the above we have used the shortand `d = dimensions` so that `d` will be a tuple of N-dimensions and then `{d: ... for d in dimensions}` such that our mapper will be valid for N-dimensional grids. Next we note the inclusion of `('middle', 1, 1)`. For mappers of the form `('middle', N, M)`, the `SubDomain` spans a contiguous region of ``dimension_size - (N + M)`` points starting at ``N`` and finishing at ``dimension_size - M``.\n",
```

Comment:
- ``dimension_size - (N + M)`` points starting at ``N`` and finishing at ``dimension_size - M``

start at `N` and finishing at ``dimension_size - M`` is `dimension_size - M-N+1` not `dimension_size - M -N`

Reply:
Indeed, good catch.